docs/codeql/codeql-language-guides/customizing-library-models-for-java.rst

```diff
@@ -0,0 +1,417 @@
+.. _customizing-library-models-for-java:
+
+:orphan:
+:nosearch:
+
+Customizing Library Models for Java
+===================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+The Java analysis can be customized by adding library models (summaries, sinks and sources) in data extension files.
+
+A data extension file for Java is a YAML file in the form:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/java-all
+         extensible: <name of extensible predicate>
+       data:
+         - <tuple1>
+         - <tuple2>
+         - ...
+
+Data extensions contribute to extensible predicates defined CodeQL libraries. For more information on how to define data extensions and extensible predicates as well as how to wire them up, see the :ref:`data-extensions` documentation.
+
+The CodeQL libraries for Java expose the following extensible predicates:
+
+- **sourceModel**\(package, type, subtypes, name, signature, ext, output, kind, provenance)
+- **sinkModel**\(package, type, subtypes, name, signature, ext, input, kind, provenance)
+- **summaryModel**\(package, type, subtypes, name, signature, ext, input, output, kind, provenance)
+- **neutralModel**\(package, type, name, signature, provenance)
+
```

Comment:
@michaelnebel are there any places in the doc we could point a user to a definition for a source, sink, etc? I see someone coming across this document at a future point trying to model a library and not being sure about what source, sink, summary or neutral model refers to.

Reply:
Good point and similar to @kaeluka 's comment. I will provide a bit more detail, but I would like to keep the scope focus on the model definitions. If it turns out that we need more extensive documentation of *taint tracking*, *sources*, *sinks*, *summaries* it should probably be documented elsewhere.